web/models/user.ex

```diff
@@ -9,22 +12,56 @@ defmodule Yggdrasil.User do
     timestamps
   end
 
-  @required_fields ~w(username password password_confirmation)
-  @optional_fields ~w()
-
   @doc """
-  Creates a changeset based on the `model` and `params`.
-
-  If no params are provided, an invalid changeset is returned
-  with no validation performed.
+  create_changeset should be used when creating a new user it ensures
+  all the fields needed are there and valid and generates a password
   """
-  def changeset(model, params \\ :empty) do
+  def create_changeset(model, params \\ :empty) do
     model
-    |> cast(params, @required_fields, @optional_fields)
+    |> cast(params, ~w(username password password_confirmation), ~w())
     |> update_change(:username, &String.downcase/1)
     |> unique_constraint(:username)
     |> validate_length(:password, min: 4)
     |> validate_length(:password_confirmation, min: 4)
     |> validate_confirmation(:password)
+    |> add_password
```

Comment:
Maybe 'hash_password' instead? There is already a password on the struct, you are hashing it and adding the hash.